### Welcome to the functorch tutorial on ensembling models, in colab.

## Configuring your colab to run functorch 


**Getting setup** - running functorch currently requires Pytorch Nightly.  
Thus we'll go through a pytorch nightly install and build functorch. 

After that and a restart, you'll be ready to run the tutorial here on colab.

Let's setup a restart function:

In [3]:
def colab_restart():
  print("--> Restarting colab instance") 
  get_ipython().kernel.do_shutdown(True)

Next, let's confirm that we have a gpu.  
(If not, select Runtime -> Change Runtime type above,
 and select GPU under Hardward Accelerator )

In [4]:
!nvcc --version

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2020 NVIDIA Corporation
Built on Mon_Oct_12_20:09:46_PDT_2020
Cuda compilation tools, release 11.1, V11.1.105
Build cuda_11.1.TC455_06.29190527_0


Let's remove the default PyTorch install:

In [5]:
!pip uninstall -y torch

Found existing installation: torch 1.11.0+cu111
Uninstalling torch-1.11.0+cu111:
  Successfully uninstalled torch-1.11.0+cu111


And install the relevant nightly version.  (this defaults to 11.1 Cuda which works on most colabs). 

In [6]:
cuda_version = "cu111" # optionally - cu113 (for 11.3) is an option as well if you have 11.3 listed above in the nvcc output. 

In [7]:
!pip install --pre torch -f https://download.pytorch.org/whl/test/{cuda_version}/torch_test.html --upgrade

Looking in links: https://download.pytorch.org/whl/test/cu111/torch_test.html
tcmalloc: large alloc 1922670592 bytes == 0x560ca6272000 @  0x7f80afd611e7 0x560ca2de4518 0x560ca2daed17 0x560ca2f2e1f5 0x560ca2ec88f8 0x560ca2db25d2 0x560ca2e9197e 0x560ca2db2229 0x560ca2ea4b3d 0x560ca2e26458 0x560ca2db39da 0x560ca2e22108 0x560ca2db39da 0x560ca2e22108 0x560ca2e211c0 0x560ca2db436c 0x560ca2db4571 0x560ca2e23633 0x560ca2e2102f 0x560ca2db436c 0x560ca2db4571 0x560ca2e23633 0x560ca2e2102f 0x560ca2db436c 0x560ca2db4571 0x560ca2e23633 0x560ca2e211c0 0x560ca2db436c 0x560ca2db4571 0x560ca2e23633 0x560ca2e2102f
tcmalloc: large alloc 1922670592 bytes == 0x560d18c0c000 @  0x7f80afd611e7 0x560ca2de4518 0x560ca2daed17 0x560ca2ec6297 0x560ca2db234c 0x560ca2ea4b3d 0x560ca2e26458 0x560ca2db39da 0x560ca2e22108 0x560ca2db39da 0x560ca2e22108 0x560ca2e211c0 0x560ca2db436c 0x560ca2db4571 0x560ca2e23633 0x560ca2e2102f 0x560ca2db436c 0x560ca2db4571 0x560ca2e23633 0x560ca2e2102f 0x560ca2db436c 0x560ca2db4571 0x560ca

Let's install Ninja to accelerate the functorch building process:

In [8]:
!pip install ninja

Next we'll install and build functorch (eta is ~6 minutes):

In [1]:
!pip install --user "git+https://github.com/pytorch/functorch.git@release/0.1"

  Cloning https://github.com/pytorch/functorch.git (to revision release/0.1) to /tmp/pip-req-build-wy2ev_56
  Running command git clone -q https://github.com/pytorch/functorch.git /tmp/pip-req-build-wy2ev_56
  Running command git checkout -b release/0.1 --track origin/release/0.1
  Switched to a new branch 'release/0.1'
  Branch 'release/0.1' set up to track remote branch 'release/0.1' from 'origin'.
  Created wheel for functorch: filename=functorch-0.1.0-cp37-cp37m-linux_x86_64.whl size=20716822 sha256=e3b16e5a3cc2a5013f35028180ea49aa14c3da199abd2c6d4cc6d92a0211a44b
  Stored in directory: /tmp/pip-ephem-wheel-cache-ac0ou12v/wheels/36/1e/b5/1f1fa47f6155cd0302354303feaf209e777785883d94956873
Successfully built functorch


Finally - restart colab and after that - just skip directly down to the '-- Tutorial Start --' section to get underway.

In [10]:
colab_restart() 

--> Restarting colab instance


## -- Tutorial Start -- 



In [2]:
# Confirm we are ready to start.  
# If this errs, please make sure you have completed the 'configuring your colab' steps above first and then return here.

import functorch    

# Model Ensembling

This example illustrates how to vectorize model ensembling, using vmap.





**What is model ensembling?**

Model ensembling combines the predictions from multiple models together. Traditionally this is done by running each model on some inputs separately and then combining the predictions. However, if you’re running models with the same architecture, then it may be possible to combine them together using vmap. vmap is a function transform that maps functions across dimensions of the input tensors. One of its use cases is eliminating for-loops and speeding them up through vectorization.

Let’s demonstrate how to do this using an ensemble of simple CNNs.



In [4]:
import torch
import torch.nn as nn
import torch.nn.functional as F
from functools import partial
torch.manual_seed(0);

In [5]:
# Here's a simple MLP
class SimpleMLP(nn.Module):
    def __init__(self):
        super(SimpleMLP, self).__init__()
        self.fc1 = nn.Linear(784, 128)
        self.fc2 = nn.Linear(128, 128)
        self.fc3 = nn.Linear(128, 10)

    def forward(self, x):
        x = x.flatten(1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.fc2(x)
        x = F.relu(x)
        x = self.fc3(x)
        return x


Let’s generate a batch of dummy data and pretend that we’re working with an MNIST dataset. Thus, the dummy images are 28 by 28, and we have a minibatch of size 64. Furthermore, lets say we want to combine the predictions from 10 different models. 


In [6]:
device = 'cuda'
num_models = 10

data = torch.randn(100, 64, 1, 28, 28, device=device)
targets = torch.randint(10, (6400,), device=device)

models = [SimpleMLP().to(device) for _ in range(num_models)]

We have a couple of options for generating predictions. Maybe we want to give each model a different randomized minibatch of data. Alternatively, maybe we want to run the same minibatch of data through each model (e.g. if we were testing the effect of different model initializations).





Option 1: different minibatch for each model

In [7]:
minibatches = data[:num_models]
predictions_diff_minibatch_loop = [model(minibatch) for model, minibatch in zip(models, minibatches)]

Option 2: Same minibatch

In [8]:
minibatch = data[0]
predictions2 = [model(minibatch) for model in models]

## Using vmap to vectorize the ensemble






Let’s use vmap to speed up the for-loop. We must first prepare the models for use with vmap.

First, let’s combine the states of the model together by stacking each parameter. For example, `model[i].fc1.weight` has shape `[784, 128]`; we are going to stack the .fc1.weight of each of the 10 models to produce a big weight of shape `[10, 784, 128]`.

functorch offers the 'combine_state_for_ensemble' convenience function to do that. It returns a stateless version of the model (fmodel) and stacked parameters and buffers.



In [9]:
from functorch import combine_state_for_ensemble

fmodel, params, buffers = combine_state_for_ensemble(models)
[p.requires_grad_() for p in params];


Option 1: get predictions using a different minibatch for each model. 

By default, vmap maps a function across the first dimension of all inputs to the passed-in function. After using the combine_state_for_ensemble, each of the params and buffers have an additional dimension of size 'num_models' at the front, and minibatches has a dimension of size 'num_models'.






In [10]:
print([p.size(0) for p in params]) # show the leading 'num_models' dimension

assert minibatches.shape == (num_models, 64, 1, 28, 28) # verify minibatch has leading dimension of size 'num_models'

[10, 10, 10, 10, 10, 10]


In [11]:
from functorch import vmap

predictions1_vmap = vmap(fmodel)(params, buffers, minibatches)

# verify the vmap predictions match the 
assert torch.allclose(predictions1_vmap, torch.stack(predictions_diff_minibatch_loop), atol=1e-3, rtol=1e-5)

Option 2: get predictions using the same minibatch of data.

vmap has an in_dims arg that specifies which dimensions to map over. By using `None`, we tell vmap we want the same minibatch to apply for all of the 10 models.




In [12]:
predictions2_vmap = vmap(fmodel, in_dims=(0, 0, None))(params, buffers, minibatch)

assert torch.allclose(predictions2_vmap, torch.stack(predictions2), atol=1e-3, rtol=1e-5)

A quick note: there are limitations around what types of functions can be transformed by vmap. The best functions to transform are ones that are pure functions: a function where the outputs are only determined by the inputs that have no side effects (e.g. mutation). vmap is unable to handle mutation of arbitrary Python data structures, but it is able to handle many in-place PyTorch operations.

## Performance

Curious about performance numbers? Here's how the numbers look on Google Colab.

In [13]:
from torch.utils.benchmark import Timer
without_vmap = Timer(
    stmt="[model(minibatch) for model, minibatch in zip(models, minibatches)]",
    globals=globals())
with_vmap = Timer(
    stmt="vmap(fmodel)(params, buffers, minibatches)",
    globals=globals())
print(f'Predictions without vmap {without_vmap.timeit(100)}')
print(f'Predictions with vmap {with_vmap.timeit(100)}')

Predictions without vmap <torch.utils.benchmark.utils.common.Measurement object at 0x7f85c3749650>
[model(minibatch) for model, minibatch in zip(models, minibatches)]
  3.20 ms
  1 measurement, 100 runs , 1 thread
Predictions with vmap <torch.utils.benchmark.utils.common.Measurement object at 0x7f85c298cd90>
vmap(fmodel)(params, buffers, minibatches)
  879.02 us
  1 measurement, 100 runs , 1 thread


There's a large speedup using vmap! 

In general, vectorization with vmap should be faster than running a function in a for-loop and competitive with manual batching. There are some exceptions though, like if we haven’t implemented the vmap rule for a particular operation or if the underlying kernels weren’t optimized for older hardware (GPUs). If you see any of these cases, please let us know by opening an issue at our [GitHub](https://github.com/pytorch/functorch)!

